### ISSUE TITLE
Compile Error When Stacking LM With Linear Layer

### ISSUE THREAD

**Issue by joelsimonoff** [aws/aws-neuron-sdk#183]
I'm having difficulties compiling a model that combines a language model with a linear layer. The model workflow is:

1. Run RoBERTa on the input
2. Apply masks to outputted vectors
3. Average the vectors after masking
3. Use a linear layer to do a feature reduction

I put together a simple scaled-down version of this model to make it easy to reproduce the issue:

```python
class DemoModel(torch.nn.Module):
    def __init__(self):
        super().__init__()

        self.lm = RobertaModel.from_pretrained("roberta-base")
        self.linear = torch.nn.Linear(768,128)
    def forward(self, toks, masks):
        out, _ = self.lm(toks, masks)
        out = torch.sum(out * masks.unsqueeze(-1), 1)
        out = out / torch.sum(masks)
        
        return self.linear(out)

tokenizer = RobertaTokenizer.from_pretrained("roberta-base")   

model = DemoModel()
model.eval()

sentence = "Some Generic example text."
inputs = tokenizer.encode_plus(sentence, return_tensors='pt', max_length=256, pad_to_max_length=True,
                               add_special_tokens=True)
input_parsed = inputs['input_ids'], inputs['attention_mask']
embeddings = model(*input_parsed) # Checks and make sure input works.
print("Embeddings Output Shape:", embeddings.shape)

model_neuron = torch.neuron.trace(model, input_parsed, compiler_args=['-O2'])
```

I get the following output:

```
Embeddings Output Shape: torch.Size([1, 128])

INFO:Neuron:Optimize = None
INFO:Neuron:Compiler args type is <class 'list'> value is ['-O2']
INFO:Neuron:compiling function _NeuronGraph$4617 with neuron-cc
INFO:Neuron:Compiling with command line: '/home/ec2-user/main/bin/neuron-cc compile /tmp/tmpvh15xs60/graph_def.pb --framework TENSORFLOW --pipeline compile SaveTemps --output /tmp/tmpvh15xs60/graph_def.neff --io-config {"inputs": {"tensor:0": [[1, 256, 768], "float32"], "1:0": [[1, 768], "float32"]}, "outputs": ["Add:0"]} -O2''

WARNING:Neuron:torch.neuron.trace failed on <torch_neuron.graph.Graph object at 0x7ff08316fed0>; falling back to native python function call

WARNING:Neuron:torch.neuron.trace error: NeuronDevice::process_nrtd_response: Context=NRTD_CTX_INFER_POST NRTD response code 'NERR_INVALID' (2) The Neuron runtime received an invalid input on this request. Raw runtime error message '[NRTD:infer_post_internal] Incorrect number of inputs, expected: 1, received: 2
[NRTD:infer_post] error: 2
'
-- Use compile_from_neff function
-- Use create_runnable function
```

This outputs a model that runs only on the CPU. However, if I modify the `forward` function in `DemoModel` such that it only returns the text model output:

```python
class DemoModel(torch.nn.Module):
    def __init__(self):
        super().__init__()

        self.lm = RobertaModel.from_pretrained("roberta-base")
        self.linear = torch.nn.Linear(768,128)
    def forward(self, toks, masks):
        out, _ = self.lm(toks, masks)
        
        return out
```

This modified `DemoModel` compiles without issue and runs on Inferentia chips. This leads me to believe that the issue is coming from somewhere in the second part of the `forward function`:

```python
out = torch.sum(out * masks.unsqueeze(-1), 1)
out = out / torch.sum(masks)
        
return self.linear(out)
```

Environment information:

- Machine: `inf1.6xlarge`
- Torch Version: `1.5.1`
- Torch-Neuron Version: `1.5.1.1.0.1721.0`
- Transformers Version: `2.8.0`

**Comment by aws-joshim** [comment]
@joelsimonoff we are working on reproducing the compile issue with the scaled down version that you shared above. We will get back to you shortly with our findings

**Comment by aws-joshim** [comment]
@joelsimonoff We have identified that the  torch.neuron.trace 'NERR_INVALID' error for this Demo model is a result of an internal error that is related to a unused input that our compiler is optimizing away. We are working on a fix for this in a subsequent release.  In the interim, the modification to the Demo model that you recommended on the ticket would be a potential workaround for this issue.

**Comment by joelsimonoff** [comment]
@aws-joshim Is it possible to compile the model in such a way where both models run on separate neuron cores and it would treat it as a pipeline? Or is the best workaround here just to run the linear layer on the CPU?

**Comment by aws-joshim** [comment]
@joelsimonoff in the short term, the optimal workaround would be to run the linear layer on the CPU. We are looking into pipelining options that could be more performant that running the linear layer on CPU. We will keep you posted if we find a better alternative.

**Comment by awsrjh** [comment]
hello @joelsimonoff --  let us know if we can help further or of this has helped solve your issue for now. 

and for reference - yes - you can assign different models (or sub-graphs of models) to execute independently on each Neuron core. In such a case, the cpu would be the connection point between those two models.

**Comment by joelsimonoff** [comment]
@awsrjh Our linear layer has input dim 768 and output dim 128. In your experience, would this run faster on a CPU than on the inf1 if you factor in the data transfer time between CPU and Inferentia? Is this something we'll have to experiment with?

**Comment by awsrjh** [comment]
yes -- this  is likely an an experimental question for now.

**Comment by awsrjh** [comment]
Joel 

has this worked out for you - can we help more?

**Comment by mrnikwaws** [comment]
Hi Joel,

I'm closing this ticket since we haven't heard back in the last two weeks.  If we can offer further assistance please re-open this ticket and comment on it.
